src/java.base/share/classes/sun/security/tools/keytool/Main.java

```diff
@@ -1922,22 +1978,56 @@ private void doGenKeyPair(String alias, String dname, String keyAlgName,
         }
 
         PrivateKey privKey = keypair.getPrivateKey();
+        CertificateExtensions ext;
 
-        CertificateExtensions ext = createV3Extensions(
-                null,
-                null,
-                v3ext,
-                keypair.getPublicKeyAnyway(),
-                null);
+        if (signerFlag) {
+            Certificate signerCert = keyStore.getCertificate(signerAlias);
+
+            X509CertImpl certImpl;
+            if (signerCert instanceof X509CertImpl) {
+                    certImpl = (X509CertImpl) signerCert;
+            } else {
+                certImpl = new X509CertImpl(signerCert.getEncoded());
+            }
+
+            KeyIdentifier signerSubjectKeyId = certImpl.getSubjectKeyId();
+            if (signerSubjectKeyId == null) {
+                signerSubjectKeyId = new KeyIdentifier(signerCert.getPublicKey());
+            }
+
+            ext = createV3Extensions(
+                    null,
+                    null,
+                    v3ext,
+                    keypair.getPublicKeyAnyway(),
+                    signerSubjectKeyId);
+        } else {
+            ext = createV3Extensions(
+                    null,
+                    null,
+                    v3ext,
+                    keypair.getPublicKeyAnyway(),
+                    null);
+        }
 
         X509Certificate[] chain = new X509Certificate[1];
         chain[0] = keypair.getSelfCertificate(
                 x500Name, getStartDate(startDate), validity*24L*60L*60L, ext);
 
-        MessageFormat form = new MessageFormat(rb.getString
-            ("Generating.keysize.bit.keyAlgName.key.pair.and.self.signed.certificate.sigAlgName.with.a.validity.of.validality.days.for"));
+        MessageFormat form;
+        if (signerFlag) {
+            form = new MessageFormat(rb.getString
+                    ("Generating.keysize.bit.keyAlgName.key.pair.and.a.certificate.sigAlgName.issued.by.an.entry.specified.by.the.signer.option.with.a.validity.of.validality.days.for"));
+        } else {
+            form = new MessageFormat(rb.getString
+                    ("Generating.keysize.bit.keyAlgName.key.pair.and.self.signed.certificate.sigAlgName.with.a.validity.of.validality.days.for"));
+            if (groupName != null) {
+                keysize = KeyUtil.getKeySize(privKey);
```

Comment:
Don't understand why `keysize` only needs to be calculated when there is no signer flag. In fact, we can just always use the `getKeySize` output below.

Reply:
For the no -signer case, calculating the `keysize` based on the `groupName` is the original logic, and `groupName` and `keysize` can not coexist. Updated the -signer case to do the same.